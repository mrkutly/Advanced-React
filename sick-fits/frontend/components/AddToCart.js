import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./User";

export const ADD_TO_CART_MUTATION = gql`
	mutation($id: ID!) {
		addToCart(id: $id) {
			id
			quantity
		}
	}
`;

class AddToCart extends React.Component {
	render() {
		const { id } = this.props.item;
		return (
			<Mutation
				mutation={ADD_TO_CART_MUTATION}
				variables={{ id }}
				refetchQueries={[{ query: CURRENT_USER_QUERY }]}
			>
				{(addToCart, { loading }) => (
					<button onClick={addToCart} disabled={loading}>
						Add{loading && "ing"} to cart
					</button>
				)}
			</Mutation>
		);
	}
}

export default AddToCart;
