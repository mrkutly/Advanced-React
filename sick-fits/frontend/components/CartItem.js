import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import formatMoney from "../lib/formatMoney";

const CartItemStyles = styled.li`
	padding: 1rem 0;
	border-bottom: 1px solid ${({ theme }) => theme.lightgrey};
	display: grid;
	align-items: center;
	grid-template-columns: auto 1fr auto;

	img {
		margin-right: 10px;
	}

	h3,
	p {
		margin: 0;
	}
`;

const CartItem = ({ cartItem: { item, quantity } }) => (
	<CartItemStyles>
		<img width="100" src={item.image} alt={item.title} />
		<div className="cart-item-details">
			<h3>{item.title}</h3>
			<p>
				{formatMoney(item.price * quantity)}
				{" - "}
				<em>
					{quantity} &times; {formatMoney(item.price)} each
				</em>
			</p>
		</div>
	</CartItemStyles>
);

CartItem.propTypes = {
	cartItem: PropTypes.object.isRequired,
};

export default CartItem;
