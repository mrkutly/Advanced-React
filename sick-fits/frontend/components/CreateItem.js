import React, { Component } from "react";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";
import Router from "next/router";
import Form from "./styles/Form";
import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";

const CREATE_ITEM_MUTATION = gql`
	# here, we tell our mutation to expect typed args
	mutation CREATE_ITEM_MUTATION(
		$title: String!
		$description: String!
		$image: String!
		$largeImage: String!
		$price: Int!
	) {
		# Now we are calling the createItem mutation from our schema.graphql in our backend,
		# using the variables passed in above. GQL variables are used like bash variables (with the $)
		createItem(
			title: $title
			description: $description
			image: $image
			largeImage: $largeImage
			price: $price
		) {
			# after the item is created, we return the id
			id
		}
	}
`;

class CreateItem extends Component {
	state = {
		title: "",
		description: "",
		image: "",
		largeImage: "",
		price: 0,
	};

	handleChange = e => {
		const { type, value, name } = e.target;
		const val = type === "number" ? parseFloat(value) : value;
		this.setState({ [name]: val });
	};

	uploadFile = async e => {
		// formatting the file to upload to cloudinary
		const files = e.target.files;
		const data = new FormData();
		data.append("file", files[0]);
		data.append("upload_preset", "sickfits");

		// upload the file to cloudinary
		const res = await fetch(
			"https://api.cloudinary.com/v1_1/mrkutly/image/upload",
			{ method: "POST", body: data }
		);

		// grab the secure_url's from the json response and set those as the image and largeImage
		const file = await res.json();
		this.setState({
			image: file.secure_url,
			largeImage: file.eager[0].secure_url,
		});
	};

	render() {
		const { title, description, image, largeImage, price } = this.state;
		// To use our CREATE_ITEM_MUTATION in our form, we wrap our entire form in a Mutation component
		// The Mutation components only child will be a function, just like the Query Components
		// This function takes in the args of the mutation and the payload (loading, error, called, and data)
		// Fun: loading is a boolean that Apollo flips on and off for us, so we can use it for our loading state
		//  Then, by wrapping all of your jsx in this function, the mutation is now exposed to your form! YAY!
		return (
			<Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
				{(createItem, { loading, error }) => (
					<Form
						onSubmit={async e => {
							// Stop form from submitting
							e.preventDefault();
							// create the item
							const res = await createItem();
							// send them to single item page
							console.log(res);
							Router.push({
								pathname: "/item",
								query: { id: res.data.createItem.id },
							});
						}}
					>
						<Error error={error} />
						<fieldset disabled={loading} aria-busy={loading}>
							<label htmlFor="file">
								Image
								<input
									type="file"
									id="file"
									name="file"
									placeholder="Upload an Image"
									onChange={this.uploadFile}
									required
								/>
							</label>
							{image && <img src={image} alt="Upload preview" width="200" />}

							<label htmlFor="title">
								Title
								<input
									type="text"
									id="title"
									name="title"
									value={title}
									placeholder="Title"
									onChange={this.handleChange}
									required
								/>
							</label>

							<label htmlFor="price">
								Price
								<input
									type="number"
									id="price"
									name="price"
									value={price}
									placeholder="Price"
									onChange={this.handleChange}
									required
								/>
							</label>

							<label htmlFor="description">
								Description
								<textarea
									type="text"
									id="description"
									name="description"
									value={description}
									placeholder="Enter a description"
									onChange={this.handleChange}
									required
								/>
							</label>
							<button type="submit">Submit</button>
						</fieldset>
					</Form>
				)}
			</Mutation>
		);
	}
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
