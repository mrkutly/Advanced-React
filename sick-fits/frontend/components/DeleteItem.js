import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import { ALL_ITEMS_QUERY } from "./Items";

const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

class DeleteItem extends Component {
	update = (cache, payload) => {
		// ? Here, we can manually update the cache on the client side.
		// ? Our mutation only makes the update on the server,
		// ? so we need to also make this change to the cache.
		//
		// 1. read the cache for the items we want using a graphql query (ALL_ITEMS_QUERY from Items.js)
		const data = cache.readQuery({ query: ALL_ITEMS_QUERY });

		// 2. Filter the deleted item out of the page
		data.items = data.items.filter(
			item => item.id !== payload.data.deleteItem.id
		);

		// 3. Put the items back!
		cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
	};

	render() {
		return (
			<Mutation
				mutation={DELETE_ITEM_MUTATION}
				variables={{ id: this.props.id }}
				update={this.update}
			>
				{(deleteItem, { error }) => (
					<button
						onClick={e => {
							if (confirm("Are you sure you want to delete this item?")) {
								deleteItem().catch(err => {
									alert(err.message);
								});
							}
						}}
					>
						{this.props.children}
					</button>
				)}
			</Mutation>
		);
	}
}

export default DeleteItem;
