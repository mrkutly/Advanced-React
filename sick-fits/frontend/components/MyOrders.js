import React from "react";
import gql from "graphql-tag";
import styled from "styled-components";
import Link from "next/link";
import { Query } from "react-apollo";
import Error from "./ErrorMessage";
import OrderItemStyles from "./styles/OrderItemStyles";
import { formatDistance } from "date-fns";
import formatMoney from "../lib/formatMoney";

const OrderUl = styled.ul`
	display: grid;
	grid-gap: 4rem;
	grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

const MY_ORDERS_QUERY = gql`
	query myOrdersQuery {
		orders(orderBy: createdAt_DESC) {
			id
			total
			createdAt
			items {
				id
				title
				price
				description
				quantity
				image
			}
		}
	}
`;

export default class MyOrders extends React.Component {
	render() {
		return (
			<Query query={MY_ORDERS_QUERY}>
				{({ data: { orders }, loading, error }) => {
					if (error) return <Error error={error} />;
					if (loading) return <p>loading...</p>;

					return (
						<div>
							<h2>You have {orders.length} orders</h2>
							<OrderUl>
								{orders.map(order => (
									<OrderItemStyles key={order.id}>
										<Link
											href={{ pathname: "/order", query: { id: order.id } }}
										>
											<a>
												<div className="order-meta">
													<p>
														{order.items.reduce(
															(acc, item) => acc + item.quantity,
															0
														)}{" "}
														items
													</p>
													<p>{order.items.length} products</p>
													<p>{formatDistance(order.createdAt, new Date())}</p>
													<p>{formatMoney(order.total)}</p>
												</div>
												<div className="images">
													{order.items.map(item => (
														<img
															src={item.image}
															alt={item.title}
															key={item.id}
														/>
													))}
												</div>
											</a>
										</Link>
									</OrderItemStyles>
								))}
							</OrderUl>
						</div>
					);
				}}
			</Query>
		);
	}
}
