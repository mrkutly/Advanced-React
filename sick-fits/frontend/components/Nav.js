import Link from "next/link";
import NavStyles from "./styles/NavStyles";
import User from "./User";
import SignOut from "./SignOut";

const Nav = () => {
	return (
		<User>
			{({ data: { me } }) => (
				<NavStyles>
					<Link href="/items">
						<a>Shop</a>
					</Link>
					{/* only show these if the person is signed in */}
					{me && (
						<>
							<Link href="/sell">
								<a>Sell</a>
							</Link>
							<Link href="/orders">
								<a>Orders</a>
							</Link>
							<Link href="/me">
								<a>Account</a>
							</Link>
							<SignOut />
						</>
					)}
					{!me && (
						<Link href="/signup">
							<a>Sign In</a>
						</Link>
					)}
				</NavStyles>
			)}
		</User>
	);
};

export default Nav;
