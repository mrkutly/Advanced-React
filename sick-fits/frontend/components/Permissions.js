import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import PropTypes from "prop-types";
import Error from "./ErrorMessage";
import Table from "./styles/Table";
import SickButton from "./styles/SickButton";
import permissionTypes from "../lib/permissionTypes";

const ALL_USERS_QUERY = gql`
	query {
		users {
			id
			name
			email
			permissions
		}
	}
`;

const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation updatePermissions($permissions: [Permission], $userId: ID!) {
		updatePermissions(permissions: $permissions, userId: $userId) {
			id
			permissions
			name
			email
		}
	}
`;

export default props => {
	return (
		<Query query={ALL_USERS_QUERY}>
			{({ data, loading, error }) =>
				console.log(data) || (
					<div>
						<Error error={error} />
						<div>
							<h2>Manage permissions</h2>
							<Table>
								<thead>
									<tr>
										<th>Name</th>
										<th>Email</th>
										{permissionTypes.map(perm => (
											<th key={perm}>{perm}</th>
										))}
										<th>&#x2193;</th>
									</tr>
								</thead>
								<tbody>
									{data.users.map(user => (
										<UserPermissions key={user.email} user={user} />
									))}
								</tbody>
							</Table>
						</div>
					</div>
				)
			}
		</Query>
	);
};

class UserPermissions extends React.Component {
	static propTypes = {
		user: PropTypes.shape({
			name: PropTypes.string,
			email: PropTypes.string,
			id: PropTypes.string,
			permissions: PropTypes.array,
		}).isRequired,
	};

	state = {
		permissions: this.props.user.permissions,
	};

	handlePermissionsChange = ({ target }) => {
		let updatedPermissions = [...this.state.permissions];

		if (target.checked) {
			updatedPermissions.push(target.value);
		} else {
			updatedPermissions = updatedPermissions.filter(
				perm => perm !== target.value
			);
		}

		this.setState({ permissions: updatedPermissions });
	};

	render() {
		const { user } = this.props;

		return (
			<Mutation
				mutation={UPDATE_PERMISSIONS_MUTATION}
				variables={{
					permissions: this.state.permissions,
					userId: this.props.user.id,
				}}
			>
				{(updatePermissions, { loading, error }) => (
					<>
						{error && <Error error={error} />}
						<tr>
							<td>{user.name}</td>
							<td>{user.email}</td>
							{permissionTypes.map(perm => (
								<td key={perm}>
									<label htmlFor={`${user.id}-permission-${perm}`}>
										<input
											id={`${user.id}-permission-${perm}`}
											type="checkbox"
											checked={this.state.permissions.includes(perm)}
											value={perm}
											onChange={this.handlePermissionsChange}
										/>
									</label>
								</td>
							))}
							<td>
								<SickButton
									type="button"
									disabled={loading}
									onClick={updatePermissions}
								>
									Updat{loading ? "ing" : "e"}
								</SickButton>
							</td>
						</tr>
					</>
				)}
			</Mutation>
		);
	}
}
