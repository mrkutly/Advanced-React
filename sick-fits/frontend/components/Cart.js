import React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import { adopt } from "react-adopt";
import User from "./User";
import TakeMyMoney from "./TakeMyMoney";
import CartItem from "./CartItem";
import CartStyles from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import SickButton from "./styles/SickButton";
import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";

// using @client directive tells apollo to look for it in local data, not on the server
export const LOCAL_STATE_QUERY = gql`
	query {
		cartOpen @client
	}
`;

export const TOGGLE_CART_MUTATION = gql`
	mutation {
		toggleCart @client
	}
`;

// This allows you to wrap your return in all three of these components
// without getting into nested render-props nightmare
const Composed = adopt({
	// you could do this, but then you get errors in the console
	// because the "children" prop is required for these components
	// user: <User />,
	user: ({ render }) => <User>{render}</User>,
	toggleCart: ({ render }) => (
		<Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>
	),
	localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
});

const Cart = props => {
	return (
		<Composed>
			{({ user, toggleCart, localState }) => {
				const { me } = user.data;
				if (!me) return null;

				return (
					<CartStyles open={localState.data.cartOpen}>
						<header>
							<CloseButton title="close" onClick={toggleCart}>
								&times;
							</CloseButton>
							<Supreme>{me.name}'s Cart</Supreme>
							<p>
								You have {me.cart.length} item
								{me.cart.length > 1 && "s"} in your cart
							</p>
						</header>
						<ul>
							{me.cart.map(cartItem => (
								<CartItem key={cartItem.id} cartItem={cartItem}>
									{cartItem.quantity}
								</CartItem>
							))}
						</ul>
						<footer>
							<p>{formatMoney(calcTotalPrice(me.cart))}</p>
							<TakeMyMoney>
								<SickButton>Checkout</SickButton>
							</TakeMyMoney>
						</footer>
					</CartStyles>
				);
			}}
		</Composed>
	);
};

export default Cart;
