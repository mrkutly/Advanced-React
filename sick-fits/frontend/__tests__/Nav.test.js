import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeUser, fakeCartItem } from "../lib/testUtils";

const notSignedInMocks = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: { data: { me: null } },
	},
];
const signedInMocks = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: { data: { me: fakeUser() } },
	},
];
const signedInMocksWithCart = [
	{
		request: { query: CURRENT_USER_QUERY },
		result: { data: { me: { ...fakeUser(), cart: [fakeCartItem()] } } },
	},
];

describe("<Nav />", () => {
	it("renders a minimal nav when the user is signed out", async () => {
		const wrapper = mount(
			<MockedProvider mocks={notSignedInMocks}>
				<Nav />
			</MockedProvider>
		);

		await wait();
		wrapper.update();

		const nav = wrapper.find('[data-test="nav-test"]');
		const navLinks = nav.find("Link");
		expect(navLinks).toHaveLength(2);
		expect(toJSON(navLinks)).toMatchSnapshot();
	});

	it("renders a full nav when the user is signed in", async () => {
		const wrapper = mount(
			<MockedProvider mocks={signedInMocks}>
				<Nav />
			</MockedProvider>
		);

		await wait();
		wrapper.update();

		const nav = wrapper.find('[data-test="nav-test"]');
		const navLinks = nav.find("Link");
		expect(navLinks).toHaveLength(4);
		expect(toJSON(navLinks)).toMatchSnapshot();
		const navLinksText = nav.find("ul").text();
		expect(navLinksText).toContain("Sign Out");
		expect(navLinksText).toContain("Shop");
		expect(navLinksText).toContain("Sell");
		expect(navLinksText).toContain("Orders");
	});

	it("renders the amount of items in the cart", async () => {
		const wrapper = mount(
			<MockedProvider mocks={signedInMocksWithCart}>
				<Nav />
			</MockedProvider>
		);

		await wait();
		wrapper.update();

		const count = wrapper.find('[data-test="nav-test"] div.count');
		expect(count.text()).toBe("3");
		expect(toJSON(count)).toMatchSnapshot();
	});
});
