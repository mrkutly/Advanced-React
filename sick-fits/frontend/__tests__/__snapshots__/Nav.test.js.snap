// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders a full nav when the user is signed in 1`] = `
Array [
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>,
  <Link
    href="/sell"
  >
    <a
      href="/sell"
      onClick={[Function]}
    >
      Sell
    </a>
  </Link>,
  <Link
    href="/orders"
  >
    <a
      href="/orders"
      onClick={[Function]}
    >
      Orders
    </a>
  </Link>,
  <Link
    href="/me"
  >
    <a
      href="/me"
      onClick={[Function]}
    >
      Account
    </a>
  </Link>,
]
`;

exports[`<Nav /> renders a minimal nav when the user is signed out 1`] = `
Array [
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
    >
      Shop
    </a>
  </Link>,
  <Link
    href="/signup"
  >
    <a
      href="/signup"
      onClick={[Function]}
    >
      Sign In
    </a>
  </Link>,
]
`;

exports[`<Nav /> renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 kDZyEW"
>
  3
</div>
`;
